hash.c

```diff
@@ -2906,8 +2923,37 @@ hash_aset_str(st_data_t *key, st_data_t *val, struct update_arg *arg, int existi
     return hash_aset(key, val, arg, existing);
 }
 
+// Like +hash_aset+, but stashes away the pre-existing value (if any) into +arg->arg+
+static int
+hash_update_value(st_data_t *key, st_data_t *val, struct update_arg *arg, int existing)
+{
+    if (existing) {
+        st_data_t pre_existing_value = *val;
+        *val = arg->arg;
+        arg->arg = pre_existing_value;
+    }
+    else {
+        *val = arg->arg;
+        arg->arg = Qundef;
+    }
```

Comment:
`Qnil`, or `Qundef`?

Reply:
OOOOhhhh this is nice. If we set `Qnil` here, then we no longer need this snippet (which just replaced `Qundef` with `Qnil`):

https://github.com/amomchilov/ruby/blob/d19bedffc0a5060245b6c663ff63e5c765509294/hash.c#L3018-L3020

And then, because we no longer access the value of `update_info` after the `tbl_update2()` call, we no longer need the `tbl_update2()`!